File: Backend/src/main/java/lan/dk/podcastserver/manager/worker/francetv/FranceTvUpdater.java
package lan.dk.podcastserver.manager.worker.francetv;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.vavr.collection.HashSet;
import io.vavr.collection.List;
import io.vavr.collection.Set;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.manager.worker.Type;
import lan.dk.podcastserver.manager.worker.Updater;
import lan.dk.podcastserver.service.*;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.jsoup.nodes.Element;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.Objects;

import static java.util.Objects.isNull;

/**
 * Created by kevin on 30/06/2017.
 */
@Slf4j
@Component("FranceTvUpdater")
@RequiredArgsConstructor
public class FranceTvUpdater implements Updater {
    private static final String LAST_VIDEOS_SELECTOR = "ul.wall";
    private static final String CATALOG_URL = "https://sivideo.webservices.francetelevisions.fr/tools/getInfosOeuvre/v2/?idDiffusion=%s";

    private final SignatureService signatureService;
    private final HtmlService htmlService;
    private final ImageService imageService;
    private final JsonService jsonService;

    public Set<Item> getItems(Podcast podcast) {
        return htmlService.get(podcast.getUrl())
                .map(p -> p.select(LAST_VIDEOS_SELECTOR))
                .flatMap(v -> HashSet.ofAll(v).find(ul -> "wall".equals(ul.className())))
                .map(ul -> ul.select("li"))
                .map(HashSet::ofAll).getOrElse(HashSet::empty)
                .map(this::htmlToItem);
    }

    private Item htmlToItem(Element element) {
        return HashSet.ofAll(element.children())
                .find(e -> Objects.equals(e.tagName(), "a"))
                .map(e -> e.attr("data-video"))
                .map(id -> String.format(CATALOG_URL, id))
                .flatMap(jsonService::parseUrl)
                .map(JsonService.to(FranceTvItem.class))
                .map(ftv -> Item.builder()
                        .title(ftv.title())
                        .description(ftv.getSynopsis())
                        .pubDate(ftv.pubDate())
                        .url(getUrl(element))
                        .cover(imageService.getCoverFromURL(ftv.getImage()))
                        .build()
                )
                .getOrElse(Item.DEFAULT_ITEM);
    }

    private String getUrl(Element element) {
        return HashSet.ofAll(element.children())
                .find(e -> Objects.equals(e.tagName(), "a"))
                .map(e -> UrlService.addProtocolIfNecessary("https:", e.attr("href")))
                .getOrElse("");
    }

    @Override
    public String signatureOf(Podcast podcast) {
        String listOfIds = htmlService.get(podcast.getUrl())
                .map(p -> p.select(LAST_VIDEOS_SELECTOR))
                .flatMap(v -> HashSet.ofAll(v).find(ul -> "wall".equals(ul.className())))
                .map(ul -> ul.select("li"))
                .map(HashSet::ofAll).getOrElse(HashSet::empty)
                .flatMap(li -> HashSet.ofAll(li.children()).find(el -> Objects.equals(el.tagName(), "a")))
                .map(e -> e.attr("data-video"))
                .toList().sorted().mkString("-");

        return signatureService.generateMD5Signature(listOfIds);
    }

    @Override
    public Type type() {
        return new Type("FranceTv", "France•tv");
    }

    @Override
    public Integer compatibility(String url) {
        return isFromFranceTv(url);
    }

    public static int isFromFranceTv(String url) {
        return StringUtils.contains(url, "www.france.tv") ? 1 : Integer.MAX_VALUE;
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private static class FranceTvItem {

        private static final ZoneId ZONE_ID = ZoneId.of("Europe/Paris");

        @Setter private String titre;
        @JsonProperty("sous_titre") @Setter private String sousTitre;
        @Setter @Getter private String synopsis;

        @Setter private String saison;
        @Setter private String episode;
        @Setter private FranceTvItem.Diffusion diffusion = new FranceTvItem.Diffusion();
        @JsonProperty("image_secure") @Getter @Setter private String image;
        @Setter @Getter private List<FranceTvItem.Video> videos = List.empty();

        String title() {
            String title = titre;

            if (StringUtils.isNotEmpty(saison)) {
                title = title + " - S" + saison;
            }

            if (StringUtils.isNotEmpty(episode)) {
                title = title + "E" + episode;
            }

            if (StringUtils.isNotEmpty(sousTitre)) {
                title = title + " - " + sousTitre;
            }

            return title;
        }

        ZonedDateTime pubDate() {
            if (isNull(diffusion.getTimestamp())) {
                return null;
            }

            return ZonedDateTime.ofInstant(Instant.ofEpochSecond(diffusion.getTimestamp()), ZONE_ID);
        }


        @JsonIgnoreProperties(ignoreUnknown = true)
        private static class Diffusion {
            @Setter @Getter private Long timestamp;
        }

        @JsonIgnoreProperties(ignoreUnknown = true)
        private static class Video {
            @Setter  private String format;
            @Setter @Getter private String url;
        }
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/worker/francetv/FranceTvUpdater.kt
package com.github.davinkevin.podcastserver.manager.worker.francetv

import arrow.core.getOrElse
import arrow.core.toOption
import arrow.syntax.collections.firstOption
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.davinkevin.podcastserver.entity.Item
import com.github.davinkevin.podcastserver.entity.Podcast
import com.github.davinkevin.podcastserver.manager.worker.*
import com.github.davinkevin.podcastserver.service.HtmlService
import com.github.davinkevin.podcastserver.service.ImageService
import com.github.davinkevin.podcastserver.service.SignatureService
import com.github.davinkevin.podcastserver.service.UrlService
import com.github.davinkevin.podcastserver.utils.k
import com.jayway.jsonpath.TypeRef
import io.vavr.collection.List
import lan.dk.podcastserver.service.JsonService
import org.apache.commons.lang3.StringUtils
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.web.util.UriComponentsBuilder
import java.net.URI
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime

/**
 * Created by kevin on 30/06/2017.
 */
@Component
class FranceTvUpdater(
        val signatureService: SignatureService,
        val htmlService: HtmlService,
        val imageService: ImageService,
        val jsonService: JsonService
) : Updater {

    private val log = LoggerFactory.getLogger(FranceTvUpdater::class.java)!!

    override fun blockingFindItems(podcast: PodcastToUpdate): Set<ItemFromUpdate> {

        val urlBuilder = UriComponentsBuilder.fromHttpUrl(podcast.url.toASCIIString())

        return htmlService
                .get(toReplayUrl(podcast.url.toASCIIString())).k()
                .map { it.select("a[href]") }
                .getOrElse {
                    log.error("No items found for podcast ${podcast.id} at ${podcast.url}, the layout may have changed")
                    listOf<Element>()
                }
                .map { urlBuilder.replacePath(it.attr("href")).toUriString() }
                .map { urlToItem(it) }
                .toSet()
    }

    private fun toReplayUrl(url: String): String = "$url/replay-videos/ajax/?page=0"

    private fun urlToItem(itemUrl: String): ItemFromUpdate =
            htmlService
                    .get(itemUrl)
                    .map { it.select("script") }
                    .getOrElse {
                        log.error("No script found for item $itemUrl")
                        Elements()
                    }
                    .map { it.html() }
                    .firstOption { it.contains("FTVPlayerVideos") }
                    .map { it.substringAfter("=").trim(';') }
                    .flatMap { jsonService.parse(it).toOption() }
                    .map { JsonService.to(PAGE_ITEM).apply(it) }
                    .getOrElse { setOf() }
                    .firstOption { it.contentId in itemUrl }
                    .map { CATALOG_URL.format(it.videoId)}
                    .flatMap { jsonService.parseUrl(it).k() }
                    .map { JsonService.to(FranceTvItem::class.java).apply(it) }
                    .map { ftv -> ItemFromUpdate(
                        title = ftv.title()!!,
                        description = ftv.synopsis!!,
                        pubDate = ftv.pubDate(),
                        url = URI(itemUrl),
                        cover = imageService.fetchCoverInformation(ftv.image)?.toCoverFromUpdate()
                    ) }
                    .getOrElse { defaultItem }

    override fun blockingSignatureOf(url: URI): String {

        val listOfIds = htmlService
                .get(toReplayUrl(url.toASCIIString())).k()
                .map { it.select("a[href]") }
                .getOrElse {
                    log.error("No items found for podcast with url $url, the layout may have changed")
                    listOf<Element>()
                }
                .map { it.attr("href") }
                .sorted()
                .joinToString("-")

        return signatureService.fromText(listOfIds)
    }

    override fun type() = Type("FranceTv", "France•tv")

    override fun compatibility(url: String?) = isFromFranceTv(url)

    companion object {
        const val CATALOG_URL = "https://sivideo.webservices.francetelevisions.fr/tools/getInfosOeuvre/v2/?idDiffusion=%s"
        internal val PAGE_ITEM = object : TypeRef<Set<FranceTvPageItem>>() {}

        fun isFromFranceTv(url: String?) =
                if (StringUtils.contains(url, "www.france.tv")) 1
                else Integer.MAX_VALUE
    }
}

@JsonIgnoreProperties(ignoreUnknown = true)
internal class FranceTvPageItem(val contentId: String, val videoId: String)

@JsonIgnoreProperties(ignoreUnknown = true)
private class FranceTvItem {

    var titre: String? = null
    var synopsis: String? = null
    var saison: String? = null
    var episode: String? = null
    var diffusion = Diffusion()
    var videos: List<Video> = List.empty()
    @JsonProperty("image_secure") var image: String? = null
    @JsonProperty("sous_titre") private val sousTitre: String? = null

    fun title(): String? {
        var title = titre

        if (!saison.isNullOrEmpty()) {
            title = "$title - S$saison"
        }

        if (!episode.isNullOrEmpty()) {
            title = "${title}E$episode"
        }

        if (!sousTitre.isNullOrEmpty()) {
            title = "$title - $sousTitre"
        }

        return title
    }

    fun pubDate(): ZonedDateTime =
            when {
                diffusion.timestamp == null -> ZonedDateTime.now()
                else -> ZonedDateTime.ofInstant(Instant.ofEpochSecond(diffusion.timestamp!!), ZONE_ID)
            }


    @JsonIgnoreProperties(ignoreUnknown = true)
    private class Diffusion {
        var timestamp: Long? = null
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private class Video {
        private val format: String? = null
        var url: String? = null
    }

    companion object {
        private val ZONE_ID = ZoneId.of("Europe/Paris")
    }
}
